src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -0,0 +1,288 @@
+            IGrouping<TextSpan, ClassifiedSpan> span)
+        {
+            // Each semantic token is represented in LSP by five numbers:
+            //     1. Token line number, relative to the previous token
+            //     2. Token start character, relative to the previous token
+            //     3. Token length
+            //     4. Token type (index) - looked up in SemanticTokensLegend.tokenTypes
+            //     5. Token modifiers - each set bit will be looked up in SemanticTokensLegend.tokenModifiers
+
+            var textSpan = span.Key;
+            var linePosition = lines.GetLinePositionSpan(textSpan).Start;
+            var lineNumber = linePosition.Line;
+            var startCharacter = linePosition.Character;
+
+            // 1. Token line number, relative to the previous token
+            var deltaLine = lineNumber - lastLineNumber;
+            data.Add(deltaLine);
+
+            // 2. Token start character, relative to the previous token
+            // (Relative to 0 or the previous token’s start if they're on the same line)
+            if (lastLineNumber == lineNumber)
+            {
+                data.Add(startCharacter - lastStartCharacter);
+            }
+            else
+            {
+                data.Add(startCharacter);
+            }
+
+            // 3. Token length
+            data.Add(textSpan.Length);
+
+            var additiveResults = span.Where(s => ClassificationTypeNames.AdditiveTypeNames.Contains(s.ClassificationType));
+
+            // 4. Token type - looked up in SemanticTokensLegend.tokenTypes.
+            var tokenTypeClassifiedSpan = span.Except(additiveResults).Single();
+            if (s_classificationTypeToSemanticTokenTypeMap.TryGetValue(tokenTypeClassifiedSpan.ClassificationType, out var tokenType))
+            {
+                var index = TokenTypes.IndexOf(t => t == tokenType);
+                if (index == -1)
+                {
+                    throw new ArgumentException($"Token type {tokenType} is not recognized.");
+                }
+
+                data.Add(index);
+            }
+            else
+            {
+                throw new NotSupportedException($"Classification type {tokenTypeClassifiedSpan.ClassificationType} is unsupported.");
+            }
+
+            // 5. Token modifiers - each set bit will be looked up in SemanticTokensLegend.tokenModifiers
+            var modifiers = 0;
+            foreach (var currentModifier in additiveResults)
+            {
+                if (s_classificationTypeToSemanticTokenModifierMap.TryGetValue(currentModifier.ClassificationType, out var modifier))
+                {
+                    var index = TokenModifiers.IndexOf(t => t == modifier);
+                    if (index == -1)
+                    {
+                        throw new ArgumentException($"Token type {modifier} is not recognized.");
+                    }
+
+                    modifiers |= index + 1;
```

Comment:
Using index and ORing it doesn't make sense to me, and adding one makes it more confusing. The enum ends up needing to be:

```
Value1
Value2
Ignored3
Value4
Ignored5
Ignored6
Ignored7
Value8
```

I think `TokenModifiers` should be a `[Flags]` enum so its clear, and the value should be used not the index.

Hopefully part of this comment makes sense, I don't know if I'm explaining myself right (or indeed, have understood the code right :P)

Reply:
+1 here, a flags enum seems appropriate